src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -25,59 +25,51 @@ public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> inst
                 .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
 
-            IEnumerable<PackageCollectionItem> pkgs = _installedPackages.Concat(pkgsWithOrigins);
-            PackageCollectionItem[] allPkgs = PerformLookup(pkgs, searchToken);
+            IPackageSearchMetadata[] intalledItems = await DoSearchAsync(_installedPackages, searchToken, cancellationToken);
+            IPackageSearchMetadata[] transitiveItems = await DoSearchAsync(pkgsWithOrigins, searchToken, cancellationToken);
+            IPackageSearchMetadata[] searchItems = intalledItems.Concat(transitiveItems).ToArray();
 
-            IEnumerable<IPackageSearchMetadata> items = await TaskCombinators.ThrottledAsync(
-                allPkgs,
-                (p, t) => GetPackageMetadataAsync(p, searchToken.SearchFilter.IncludePrerelease, t),
-                cancellationToken);
-
-            return CreateResult(items);
+            return CreateResult(searchItems);
         }
 
-        internal async Task<IPackageSearchMetadata> GetPackageMetadataAsync(PackageCollectionItem identity, bool includePrerelease, CancellationToken cancellationToken)
+        internal override async Task<IPackageSearchMetadata> GetPackageMetadataAsync<T>(T pkgIdentity, bool includePrerelease, CancellationToken cancellationToken)
         {
-            IEnumerable<ITransitivePackageReferenceContextInfo> transitivePRs = identity.PackageReferences.OfType<ITransitivePackageReferenceContextInfo>();
-
-            ITransitivePackageReferenceContextInfo transitivePR = transitivePRs.OrderByDescending(x => x.Identity.Version).FirstOrDefault();
+            cancellationToken.ThrowIfCancellationRequested();
 
-            if (transitivePR != null)
+            if (pkgIdentity is PackageCollectionItem identity)
             {
-                IReadOnlyCollection<PackageIdentity> transitiveOrigins = transitivePR.TransitiveOrigins?.Select(to => to.Identity).ToArray() ?? Array.Empty<PackageIdentity>();
+                IEnumerable<ITransitivePackageReferenceContextInfo> transitivePRs = identity.PackageReferences.OfType<ITransitivePackageReferenceContextInfo>();
 
-                if (transitiveOrigins.Any())
+                ITransitivePackageReferenceContextInfo transitivePR = transitivePRs.OrderByDescending(x => x.Identity.Version).FirstOrDefault();
+
+                if (transitivePR != null)
                 {
-                    // Get only local metadata. We don't want Deprecation and Vulnerabilities Metadata on Transitive packages
-                    IPackageSearchMetadata packageMetadata = await _metadataProvider.GetOnlyLocalPackageMetadataAsync(identity, cancellationToken);
+                    IReadOnlyCollection<PackageIdentity> transitiveOrigins = transitivePR.TransitiveOrigins?.Select(to => to.Identity).ToArray() ?? Array.Empty<PackageIdentity>();
```

Comment:
```suggestion
                    IReadOnlyCollection<PackageIdentity> transitiveOrigins = transitivePR?.TransitiveOrigins?.Select(to => to.Identity).ToArray() ?? Array.Empty<PackageIdentity>();
```

Then you can avoid the outer if to check if transitivePR is null altogether because transitiveOrigins.Any will be empty.